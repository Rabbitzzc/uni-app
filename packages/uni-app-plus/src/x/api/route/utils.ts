import { ComponentPublicInstance } from 'vue'
import { removePage } from '../../../service/framework/page/getCurrentPages'
import { closeWebview } from './webview'
import { removeTabBarPage } from '../../framework/app/tabBar'

export function closePage(
  page: ComponentPublicInstance,
  animationType: string,
  animationDuration?: number
) {
  closeWebview(page.$nativePage!, animationType, animationDuration)
  removePage(page)
  removeTabBarPage(page)
}
