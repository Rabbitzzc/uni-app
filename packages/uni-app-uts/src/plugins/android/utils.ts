import fs from 'fs'
import path from 'path'
import { init, parse } from 'es-module-lexer'
import {
  normalizeNodeModules,
  normalizePath,
  parseUniExtApiNamespacesJsOnce,
  removeExt,
} from '@dcloudio/uni-cli-shared'
import {
  camelize,
  capitalize,
  isArray,
  isPlainObject,
  isString,
} from '@vue/shared'

export const UVUE_CLASS_NAME_PREFIX = 'Gen'

export const DEFAULT_APPID = 'HBuilder'

export const ENTRY_FILENAME = 'index.uts'

export async function parseImports(code: string) {
  await init
  const [imports] = parse(code)
  return imports
    .map(({ s, e }) => {
      return `import "${code.slice(s, e)}"`
    })
    .concat(parseUniExtApiImports(code))
    .join('\n')
}

function parseUniExtApiImports(code: string): string[] {
  const extApis = parseUniExtApiNamespacesJsOnce(
    process.env.UNI_UTS_PLATFORM,
    process.env.UNI_UTS_TARGET_LANGUAGE
  )
  const pattern = /uni\.(\w+)/g
  const apis = new Set<string>()
  let match
  while ((match = pattern.exec(code)) !== null) {
    apis.add(match[1])
  }
  const imports: string[] = []
  apis.forEach((api) => {
    const extApi = extApis[api]
    if (extApi) {
      imports.push(`import "${extApi[0]}"`)
    }
  })
  return imports
}

export function kotlinOutDir() {
  return path.join(process.env.UNI_OUTPUT_DIR, '../.kotlin')
}

export function uvueOutDir() {
  return path.join(process.env.UNI_OUTPUT_DIR, '../.uvue')
}

export function isVue(filename: string) {
  return filename.endsWith('.vue') || filename.endsWith('.uvue')
}

export function stringifyMap(obj: unknown, ts = false) {
  return serialize(obj, ts)
}

function serialize(obj: unknown, ts = false): string {
  if (isString(obj)) {
    return `"${obj}"`
  } else if (isPlainObject(obj)) {
    const entries = Object.entries(obj).map(
      ([key, value]) => `[${serialize(key, ts)},${serialize(value, ts)}]`
    )
    if (entries.length) {
      return `utsMapOf([${entries.join(',')}])`
    }
    if (ts) {
      return `utsMapOf<string, any | null>()`
    }
    return `utsMapOf()`
  } else if (isArray(obj)) {
    return `[${obj.map((item) => serialize(item, ts)).join(',')}]`
  } else {
    return String(obj)
  }
}

export function parseUTSRelativeFilename(filename: string) {
  if (!path.isAbsolute(filename)) {
    return filename
  }
  return normalizeNodeModules(
    path.relative(process.env.UNI_INPUT_DIR, filename)
  )
}

export function parseUTSImportFilename(filename: string) {
  if (!path.isAbsolute(filename)) {
    return filename
  }
  return normalizePath(
    path.join(
      uvueOutDir(),
      normalizeNodeModules(path.relative(process.env.UNI_INPUT_DIR, filename))
    )
  )
}

type UniCloudSpace = {
  provider: string
  spaceName: string
  spaceId: string
  clientSecret?: string
  endpoint?: string
  workspaceFolder?: string
}

let uniCloudSpaceList: Array<UniCloudSpace>
export function getUniCloudSpaceList(): Array<UniCloudSpace> {
  if (uniCloudSpaceList) {
    return uniCloudSpaceList
  }
  if (!process.env.UNI_CLOUD_SPACES) {
    uniCloudSpaceList = []
    return uniCloudSpaceList
  }
  try {
    const spaces = JSON.parse(process.env.UNI_CLOUD_SPACES)
    if (!Array.isArray(spaces)) {
      uniCloudSpaceList = []
      return uniCloudSpaceList
    }
    uniCloudSpaceList = spaces.map((space) => {
      if (space.provider === 'tcb') {
        space.provider = 'tencent'
      }
      if (space.clientSecret) {
        return {
          provider: space.provider,
          spaceName: space.name,
          spaceId: space.id,
          clientSecret: space.clientSecret,
          endpoint: space.apiEndpoint,
          workspaceFolder: space.workspaceFolder,
        }
      } else {
        return {
          provider: space.provider,
          spaceName: space.name,
          spaceId: space.id,
          workspaceFolder: space.workspaceFolder,
        }
      }
    })
  } catch (e) {
    console.error(e)
  }
  uniCloudSpaceList = uniCloudSpaceList || []
  if (uniCloudSpaceList.length > 1) {
    console.warn('Multi uniCloud space is not supported yet.')
  }
  return uniCloudSpaceList
}

// TODO copy from uni-uts-v1 compiler, should be refactor
type UniCloudObjectInfo = {
  name: string
  methodList: string[]
}

export function getUniCloudObjectInfo(
  uniCloudSpaceList: Array<UniCloudSpace>
): Array<UniCloudObjectInfo> {
  let uniCloudWorkspaceFolder = process.env.UNI_INPUT_DIR.endsWith('src')
    ? path.resolve(process.env.UNI_INPUT_DIR, '..')
    : process.env.UNI_INPUT_DIR
  let serviceProvider = 'aliyun'
  if (uniCloudSpaceList && uniCloudSpaceList.length > 0) {
    const space = uniCloudSpaceList[0]
    if (space.workspaceFolder) {
      uniCloudWorkspaceFolder = space.workspaceFolder
    }
    serviceProvider = space.provider === 'tencent' ? 'tcb' : space.provider
  } else {
    serviceProvider =
      ['aliyun', 'tcb', 'alipay'].find((item) =>
        fs.existsSync(path.resolve(uniCloudWorkspaceFolder, 'uniCloud-' + item))
      ) || 'aliyun'
  }
  try {
    const { getWorkspaceObjectInfo } = require('../../../lib/unicloud-utils')
    return getWorkspaceObjectInfo(uniCloudWorkspaceFolder, serviceProvider)
  } catch (e) {
    console.error(e)
    return []
  }
}

const extApiComponents: Set<string> = new Set()
export function addExtApiComponents(components: string[]) {
  components.forEach((component) => {
    extApiComponents.add(component)
  })
}

export function getExtApiComponents() {
  return extApiComponents
}

export function genClassName(fileName: string, prefix: string = 'Gen') {
  return (
    prefix +
    capitalize(
      camelize(
        verifySymbol(
          removeExt(
            normalizeNodeModules(fileName)
              .replace(/[\/|_]/g, '-')
              .replace(/-+/g, '-')
          )
        )
      )
    )
  )
}

function isValidStart(c: string): boolean {
  return !!c.match(/^[A-Za-z_-]$/)
}

function isValidContinue(c: string): boolean {
  return !!c.match(/^[A-Za-z0-9_-]$/)
}

function verifySymbol(s: string) {
  const chars = Array.from(s)

  if (isValidStart(chars[0]) && chars.slice(1).every(isValidContinue)) {
    return s
  }

  const buf: string[] = []
  let hasStart = false

  for (const c of chars) {
    if (!hasStart && isValidStart(c)) {
      hasStart = true
      buf.push(c)
    } else if (isValidContinue(c)) {
      buf.push(c)
    }
  }

  if (buf.length === 0) {
    buf.push('_')
  }

  return buf.join('')
}
