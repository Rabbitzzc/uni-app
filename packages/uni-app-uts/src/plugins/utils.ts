import path from 'node:path'
import { isAppNVueNativeTag } from '@dcloudio/uni-shared'
import {
  MANIFEST_JSON_UTS,
  PAGES_JSON_UTS,
  UniViteCopyPluginOptions,
  UniVitePlugin,
  initI18nOptions,
  injectAssetPlugin,
  matchUTSComponent,
  normalizeNodeModules,
  normalizePath,
  transformTapToClick,
  transformUTSComponent,
} from '@dcloudio/uni-cli-shared'
import { compileI18nJsonStr } from '@dcloudio/uni-i18n'
import { Plugin, ResolvedConfig } from 'vite'

export function createUniOptions(
  platform: 'android' | 'ios'
): UniVitePlugin['uni'] {
  return {
    copyOptions() {
      const platform = process.env.UNI_PLATFORM
      const inputDir = process.env.UNI_INPUT_DIR
      const outputDir = process.env.UNI_OUTPUT_DIR
      const targets: UniViteCopyPluginOptions['targets'] = []
      // 自动化测试时，不启用隐私政策
      if (!process.env.UNI_AUTOMATOR_WS_ENDPOINT) {
        if (process.env.UNI_UTS_PLATFORM === 'app-android') {
          targets.push({
            src: 'androidPrivacy.json',
            dest: outputDir,
            transform(source) {
              const options = initI18nOptions(platform, inputDir, false, true)
              if (!options) {
                return
              }
              return compileI18nJsonStr(source.toString(), options)
            },
          })
        }
      }
      return {
        assets: ['hybrid/html/**/*', 'uni_modules/*/hybrid/html/**/*'],
        targets,
      }
    },
    compilerOptions:
      platform === 'ios'
        ? {
            isNativeTag(tag) {
              return matchUTSComponent(tag) || isAppNVueNativeTag(tag)
            },
            nodeTransforms: [transformTapToClick, transformUTSComponent],
          }
        : {},
  }
}

export function isManifest(id: string) {
  return id.endsWith(MANIFEST_JSON_UTS)
}

export function isPages(id: string) {
  return id.endsWith(PAGES_JSON_UTS)
}

// TODO vite 升级需要考虑调整以下列表
const REMOVED_PLUGINS = [
  'vite:build-metadata',
  'vite:modulepreload-polyfill',
  // 'vite:css', // iOS replace
  'vite:esbuild',
  'vite:wasm-helper',
  'vite:worker',
  // 'vite:json',
  // 'vite:asset', // replace
  'vite:wasm-fallback',
  'vite:define',
  // 'vite:css-post', // iOS replace
  'vite:build-html',
  'vite:html-inline-proxy',
  'vite:worker-import-meta-url',
  'vite:asset-import-meta-url',
  'vite:force-systemjs-wrap-complete',
  'vite:watch-package-data',
  'commonjs',
  'vite:data-uri',
  'vite:dynamic-import-vars',
  'vite:import-glob',
  'vite:build-import-analysis',
  'vite:esbuild-transpile',
  'vite:terser',
  'vite:reporter',
]

export function configResolved(config: ResolvedConfig, isAndroidX = false) {
  const plugins = config.plugins as Plugin[]
  const len = plugins.length
  const removedPlugins = REMOVED_PLUGINS.slice(0)
  if (isAndroidX) {
    removedPlugins.push('vite:css')
    removedPlugins.push('vite:css-post')
  }
  for (let i = len - 1; i >= 0; i--) {
    const plugin = plugins[i]
    if (removedPlugins.includes(plugin.name)) {
      plugins.splice(i, 1)
    }
  }
  // console.log(plugins.map((p) => p.name))
  // 强制不inline
  config.build.assetsInlineLimit = 0
  injectAssetPlugin(config, { isAndroidX })
}

export function relativeInputDir(filename: string) {
  const inputDir = process.env.UNI_INPUT_DIR
  filename = normalizeNodeModules(filename)
  if (filename.startsWith(inputDir)) {
    return normalizePath(path.relative(inputDir, filename))
  }
  return filename
}
