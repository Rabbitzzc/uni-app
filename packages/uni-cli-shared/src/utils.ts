import fs from 'fs'
import os from 'os'
import path from 'path'
import colors from 'picocolors'
import { camelize, capitalize } from '@vue/shared'
export { default as hash } from 'hash-sum'
import {
  EXTNAME_TS_RE,
  PAGE_EXTNAME,
  PAGE_EXTNAME_APP,
  X_PAGE_EXTNAME,
  X_PAGE_EXTNAME_APP,
} from './constants'

import {
  NodeTypes,
  ElementNode,
  RootNode,
  TemplateChildNode,
} from '@vue/compiler-core'
import { ParserPlugin } from '@babel/parser'
import { getPlatformDir } from './platform'

export const version = require('../package.json').version

// 专为 uts.ts 服务
export { camelize, capitalize, isArray } from '@vue/shared'

export let isRunningWithYarnPnp: boolean
try {
  isRunningWithYarnPnp = Boolean(require('pnpapi'))
} catch {}

export const isWindows = os.platform() === 'win32'
export function normalizePath(id: string): string {
  return isWindows ? id.replace(/\\/g, '/') : id
}

export function checkElementNodeTag(
  node: RootNode | TemplateChildNode | null | undefined,
  tag: string
): node is ElementNode {
  return !!node && node.type === NodeTypes.ELEMENT && node.tag === tag
}

export function normalizeIdentifier(str: string) {
  return capitalize(camelize(str.replace(/\//g, '-')))
}

export function normalizePagePath(pagePath: string, platform: UniApp.PLATFORM) {
  const absolutePagePath = path.resolve(process.env.UNI_INPUT_DIR, pagePath)
  const isX = process.env.UNI_APP_X === 'true'
  let extensions = isX ? X_PAGE_EXTNAME : PAGE_EXTNAME
  if (platform === 'app') {
    extensions = isX ? X_PAGE_EXTNAME_APP : PAGE_EXTNAME_APP
  }
  for (let i = 0; i < extensions.length; i++) {
    const extname = extensions[i]
    if (fs.existsSync(absolutePagePath + extname)) {
      return pagePath + extname
    }
  }
  console.error(`${pagePath} not found`)
}

export function removeExt(str: string) {
  return str.split('?')[0].replace(/\.\w+$/g, '')
}

const NODE_MODULES_REGEX = /(\.\.\/)?node_modules/g

export function normalizeNodeModules(str: string) {
  str = normalizePath(str).replace(NODE_MODULES_REGEX, 'node-modules')
  // HBuilderX 内置模块路径转换
  str = str.replace(
    /.*\/plugins\/uniapp-cli-vite\/node[-_]modules/,
    'node-modules'
  )
  if (process.env.UNI_PLATFORM === 'mp-alipay') {
    str = str.replace('node-modules/@', 'node-modules/npm-scope-')
  }
  return str
}

export function normalizeMiniProgramFilename(
  filename: string,
  inputDir?: string
) {
  if (!inputDir || !path.isAbsolute(filename)) {
    return normalizeNodeModules(filename)
  }
  return normalizeNodeModules(path.relative(inputDir, filename))
}

export function normalizeParsePlugins(
  importer: string,
  babelParserPlugins?: ParserPlugin[]
) {
  const isTS = EXTNAME_TS_RE.test(importer.split('?')[0])
  const plugins: ParserPlugin[] = []
  if (isTS) {
    plugins.push('jsx')
  }
  if (babelParserPlugins) plugins.push(...babelParserPlugins)
  if (isTS) plugins.push('typescript', 'decorators-legacy')
  return plugins
}

export function pathToGlob(
  pathString: string,
  glob: string,
  options: { windows?: boolean; escape?: boolean } = {}
): string {
  const isWindows =
    'windows' in options ? options.windows : /^win/.test(process.platform)
  const useEscape = options.escape
  const str = isWindows ? pathString.replace(/\\/g, '/') : pathString
  let safeStr = str.replace(
    /[\\*?[\]{}()!]/g,
    isWindows || !useEscape ? '[$&]' : '\\$&'
  )
  return path.posix.join(safeStr, glob)
}

export function resolveSourceMapPath(
  outputDir?: string,
  platform?: UniApp.PLATFORM
) {
  return path.resolve(
    outputDir || process.env.UNI_OUTPUT_DIR,
    '../.sourcemap/' + (platform || getPlatformDir())
  )
}

function hasProjectYarn(cwd: string) {
  return fs.existsSync(path.join(cwd, 'yarn.lock'))
}

function hasProjectPnpm(cwd: string) {
  return fs.existsSync(path.join(cwd, 'pnpm-lock.yaml'))
}

function getInstallCommand(cwd: string) {
  return hasProjectYarn(cwd)
    ? 'yarn add'
    : hasProjectPnpm(cwd)
    ? 'pnpm i'
    : 'npm i'
}

export function installDepTips(
  type: 'dependencies' | 'devDependencies',
  module: string,
  version?: string
) {
  return `Cannot find module: ${module}
Please run \`${colors.cyan(
    `${getInstallCommand(process.cwd())} ${
      module + (version ? '@' + version : '')
    }${type === 'devDependencies' ? ' -D' : ''}`
  )}\` and try again.`
}

export function isAppVue(filename: string) {
  return filename.endsWith('App.vue') || filename.endsWith('App.uvue')
}

export function resolveAppVue(inputDir: string) {
  const appUVue = path.resolve(inputDir, 'App.uvue')
  if (fs.existsSync(appUVue)) {
    return normalizePath(appUVue)
  }
  return normalizePath(path.resolve(inputDir, 'App.vue'))
}
