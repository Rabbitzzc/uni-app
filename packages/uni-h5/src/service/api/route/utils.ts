import { EventChannel, parseUrl } from '@dcloudio/uni-shared'
import { type Router, isNavigationFailure } from 'vue-router'
import { createPageState } from '../../../framework/setup/page'

export type NavigateType =
  | 'navigateTo'
  | 'redirectTo'
  | 'reLaunch'
  | 'switchTab'

interface NavigateOptions {
  type: NavigateType
  url: string
  tabBarText?: string
  events?: Record<string, any>
  isAutomatedTesting?: boolean
}
export function navigate(
  { type, url, tabBarText, events, isAutomatedTesting }: NavigateOptions,
  __id__?: number
): Promise<void | { eventChannel?: EventChannel; __id__?: number }> {
  const router = getApp().$router as Router
  const { path, query } = parseUrl(url)
  return new Promise((resolve, reject) => {
    const state = createPageState(type, __id__)
    router[type === 'navigateTo' ? 'push' : 'replace']({
      path,
      query,
      state,
      force: true,
    }).then((failure) => {
      if (isNavigationFailure(failure)) {
        return reject(failure.message)
      }
      if (type === 'switchTab') {
        router.currentRoute.value.meta.tabBarText = tabBarText
      }
      if (type === 'navigateTo') {
        const meta = router.currentRoute.value.meta
        // if getOpenerEventChannel is called before navigateTo
        if (!meta.eventChannel) {
          meta.eventChannel = new EventChannel(state.__id__, events)
        } else if (events) {
          Object.keys(events).forEach((eventName) => {
            ;(meta.eventChannel as EventChannel)._addListener(
              eventName,
              'on',
              events[eventName]
            )
          })
          ;(meta.eventChannel as EventChannel)._clearCache()
        }
        return isAutomatedTesting
          ? resolve({
              __id__: state.__id__,
            })
          : resolve({
              eventChannel: meta.eventChannel as EventChannel,
            })
      }
      return isAutomatedTesting ? resolve({ __id__: state.__id__ }) : resolve()
    })
  })
}
