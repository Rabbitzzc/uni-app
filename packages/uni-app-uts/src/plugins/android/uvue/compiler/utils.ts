import path from 'path'
import { type Node } from '@babel/types'
import { camelize, capitalize } from '@vue/shared'
import { ExpressionNode, createSimpleExpression } from '@vue/compiler-core'
import { MagicString, walk } from '@vue/compiler-sfc'
import { parseExpression } from '@babel/parser'
import { normalizePath } from '@dcloudio/uni-cli-shared'
import { CompilerOptions } from './options'
import { stringifyExpression } from './transforms/transformExpression'
import { TransformContext } from './transform'
import { CompilerError } from './errors'
import { genClassName } from '../../utils'

export function genRenderFunctionDecl({ className }: CompilerOptions): string {
  return `function ${className}Render(): VNode | null`
}

export function rewriteObjectExpression(
  exp: ExpressionNode,
  context: TransformContext
) {
  const source = stringifyExpression(exp)
  if (source.includes('{')) {
    const s = new MagicString(source)
    const ast = parseExpression(source, {
      plugins: context.expressionPlugins,
    })
    walk(ast, {
      enter(node: Node) {
        if (node.type === 'ObjectExpression') {
          s.prependLeft(
            node.start!,
            node.properties.length > 0
              ? 'utsMapOf('
              : 'utsMapOf<string, any | null>('
          )
          s.prependRight(node.end!, ')')
        }
      },
    })
    return createSimpleExpression(s.toString(), false, exp.loc)
  }
}

export function onCompilerError(error: CompilerError) {}

export function genImportComponentPublicInstance(
  rootDir: string,
  tagName: string,
  fileName: string
) {
  if (fileName.includes('@dcloudio')) {
    return ''
  }
  return `, { ${genComponentPublicInstanceImported(
    rootDir,
    fileName
  )} as ${genComponentPublicInstanceIdent(tagName)} }`
}

export function genComponentPublicInstanceIdent(tagName: string) {
  return capitalize(camelize(tagName)) + 'ComponentPublicInstance'
}

export function genComponentPublicInstanceImported(
  root: string,
  fileName: string
) {
  if (path.isAbsolute(fileName)) {
    fileName = normalizePath(path.relative(root, fileName))
  }
  if (fileName.startsWith('@/')) {
    return genClassName(fileName.replace('@/', '')) + 'ComponentPublicInstance'
  }
  return genClassName(fileName) + 'ComponentPublicInstance'
}

export function addEasyComponentAutoImports(
  easyComponentAutoImports: Record<string, [string, string]>,
  rootDir: string,
  tagName: string,
  fileName: string
) {
  // 内置easycom，如 unicloud-db
  if (fileName.includes('@dcloudio')) {
    return
  }
  if (path.isAbsolute(fileName)) {
    fileName = '@/' + normalizePath(path.relative(rootDir, fileName))
  }
  easyComponentAutoImports[fileName] = [
    genComponentPublicInstanceImported(rootDir, fileName),
    genComponentPublicInstanceIdent(tagName),
  ]
}
