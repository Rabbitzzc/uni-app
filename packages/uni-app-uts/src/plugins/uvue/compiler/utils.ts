import { type Node } from '@babel/types'
import { ExpressionNode, createSimpleExpression } from '@vue/compiler-core'
import { MagicString, walk } from '@vue/compiler-sfc'
import { parseExpression } from '@babel/parser'
import { CompilerOptions } from './options'

import { stringifyExpression } from './transforms/transformExpression'

import { TransformContext } from './transform'
import { CompilerError } from './errors'
import { camelize, capitalize } from '@vue/shared'
import { normalizeNodeModules, removeExt } from '@dcloudio/uni-cli-shared'
import path from 'path'
import { UVUE_CLASS_NAME_PREFIX } from '../../utils'

export function genRenderFunctionDecl({
  targetLanguage,
  className,
}: CompilerOptions): string {
  return `${
    targetLanguage === 'kotlin' ? '@Suppress("UNUSED_PARAMETER") ' : ''
  }function ${className}Render(): VNode | null`
}

export function rewriteObjectExpression(
  exp: ExpressionNode,
  context: TransformContext
) {
  const source = stringifyExpression(exp)
  if (source.includes('{')) {
    const s = new MagicString(source)
    const ast = parseExpression(source, {
      plugins: context.expressionPlugins,
    })
    walk(ast, {
      enter(node: Node) {
        if (node.type === 'ObjectExpression') {
          s.prependLeft(
            node.start!,
            node.properties.length > 0
              ? 'utsMapOf('
              : 'utsMapOf<string, any | null>('
          )
          s.prependRight(node.end!, ')')
        }
      },
    })
    return createSimpleExpression(s.toString(), false, exp.loc)
  }
}

export function onCompilerError(error: CompilerError) {}

export function genImportComponentPublicInstance(
  rootDir: string,
  tagName: string,
  fileName: string
) {
  if (fileName.includes('@dcloudio')) {
    return ''
  }
  return `, { ${genComponentPublicInstanceImported(
    rootDir,
    fileName
  )} as ${genComponentPublicInstanceIdent(tagName)} }`
}

function genComponentPublicInstanceIdent(tagName: string) {
  return capitalize(camelize(tagName)) + 'ComponentPublicInstance'
}

export function genComponentPublicInstanceImported(
  root: string,
  fileName: string
) {
  if (fileName.startsWith('@/')) {
    return (
      UVUE_CLASS_NAME_PREFIX +
      capitalize(
        camelize(
          removeExt(fileName.replace('@/', ''))
            .replace(/\//g, '-')
            .replace(/@/g, '')
            .replace(/\./g, '')
        )
      ) +
      'ComponentPublicInstance'
    )
  }
  return (
    UVUE_CLASS_NAME_PREFIX +
    capitalize(
      camelize(
        removeExt(normalizeNodeModules(path.relative(root, fileName)))
          .replace(/\//g, '-')
          .replace(/@/g, '')
          .replace(/\./g, '')
      )
    ) +
    'ComponentPublicInstance'
  )
}
