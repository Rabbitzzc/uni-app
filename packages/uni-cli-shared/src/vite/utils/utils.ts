import type { ConfigEnv, ResolvedConfig, UserConfig } from 'vite'
import type { RollupError } from 'rollup'
import type { CompilerError } from '@vue/compiler-sfc'
import { extend } from '@vue/shared'
import { codeFrameColumns } from '@babel/code-frame'
import { offsetToStartAndEnd } from '../plugins/vitejs/utils'

export function withSourcemap(config: ResolvedConfig) {
  if (config.command === 'serve') {
    return true
  }
  return !!config.build.sourcemap
}

export function isInHybridNVue(config: UserConfig | ResolvedConfig): boolean {
  return (config as any).nvue && process.env.UNI_RENDERER !== 'native'
}

export function isSsr(
  command: ConfigEnv['command'],
  config: UserConfig | ResolvedConfig
) {
  if (command === 'serve') {
    return !!(config.server && config.server.middlewareMode)
  }
  if (command === 'build') {
    return !!(config.build && config.build.ssr)
  }
  return false
}

export function createRollupError(
  plugin: string,
  id: string,
  error: CompilerError | SyntaxError,
  source?: string
): RollupError {
  const { message, name, stack } = error
  const rollupError: RollupError = extend(new Error(message), {
    id,
    plugin,
    name,
    stack,
  })

  if ('code' in error && error.loc) {
    rollupError.loc = {
      file: id,
      line: error.loc.start.line,
      column: error.loc.start.column,
    }
    if (source && source.length > 0) {
      if ('offsetStart' in error && 'offsetEnd' in error) {
        rollupError.frame = codeFrameColumns(
          source,
          offsetToStartAndEnd(
            source,
            error.offsetStart as number,
            error.offsetEnd as number
          )
        )
      } else {
        rollupError.frame = codeFrameColumns(source, error.loc)
      }
    }
  }
  if (id) {
    // 指定了id后，不让后续的rollup重写
    Object.defineProperty(rollupError, 'id', {
      get() {
        return id
      },
      set(_v) {},
    })
  }
  return rollupError
}

export const generateCodeFrameColumns = codeFrameColumns
