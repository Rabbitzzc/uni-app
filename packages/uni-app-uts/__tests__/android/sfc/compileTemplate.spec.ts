import { compileSFCScript as compile, assertCode } from './utils'

describe('SFC compile template', () => {
  test('template with html', () => {
    const { content } = compile(
      `<script setup>const msg = ''</script><template lang='html'><view class="test"/></template>`,
      { inlineTemplate: true }
    )
    assertCode(content)
  })
  test('template with pug', () => {
    const { content } = compile(
      `<script setup>const msg = ''</script><template lang='pug'>view.test</template>`,
      { inlineTemplate: true }
    )
    assertCode(content)
  })
})
