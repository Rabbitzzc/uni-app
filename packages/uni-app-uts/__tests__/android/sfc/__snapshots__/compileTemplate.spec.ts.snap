// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SFC compile template template with html 1`] = `
"const msg = ''
export default {
  setup(__props: GenAnonymous) {
const __ins = getCurrentInstance()!;
const _ctx = __ins.proxy as GenAnonymous;
const _cache = __ins.renderCache;

return 
function GenAnonymousRender(): any | null {

  return createElementVNode("view", utsMapOf({ class: "test" }))
}
}

}"
`;

exports[`SFC compile template template with pug 1`] = `
"const msg = ''
export default {
  setup(__props: GenAnonymous) {
const __ins = getCurrentInstance()!;
const _ctx = __ins.proxy as GenAnonymous;
const _cache = __ins.renderCache;

return 
function GenAnonymousRender(): any | null {

  return createElementVNode("view", utsMapOf({ class: "test" }))
}
}

}"
`;
